File: ignite-nest/src/post/service/post-detail.service.ts
import { Inject, Injectable } from '@nestjs/common';
import { POST_IN_PORT } from '../constants/post.token';
import { PostInPort } from '../domain/port/in/post-in.port';
import { PostDetailUsecase } from '../domain/usecase/post-detail.usecase';
import { COMMENT_IN_PORT } from '../../comment/constants/comment.token';
import { CommentInPort } from '../../comment/domain/port/in/comment-in.port';
import { PostDetailDto } from '../infrastructure/web/dto/response/post-detail.dto';
import { CommentMapper } from '../../comment/mapper/comment.mapper';
import { PostMapper } from '../mapper/post.mapper';

@Injectable()
export class PostDetailService implements PostDetailUsecase {
  constructor(
    @Inject(POST_IN_PORT)
    private readonly postInPort: PostInPort,
    @Inject(COMMENT_IN_PORT)
    private readonly commentInPort: CommentInPort,
  ) {}

  async execute(postId: number): Promise<PostDetailDto> {
    const post = await this.postInPort.getPostByIdOrThrow(postId);
    const comments = (
      await this.commentInPort.findCommentsByPostId(postId)
    ).map((comment) => {
      return CommentMapper.toPostDetailResponse(comment);
    });
    const response = PostMapper.toResponse(post, comments);
    return response;
  }
}

File: ignite-spring/src/main/kotlin/org/rookedsysc/ignitespring/post/application/service/PostDetailService.kt
package org.rookedsysc.ignitespring.post.application.service

import org.rookedsysc.ignitespring.comment.domain.port.`in`.CommentInPort
import org.rookedsysc.ignitespring.comment.entity.CommentEntity
import org.rookedsysc.ignitespring.comment.infrastructure.web.dto.CommentPostDetailListDto
import org.rookedsysc.ignitespring.comment.mapper.CommentMapper
import org.rookedsysc.ignitespring.post.domain.port.`in`.PostInPort
import org.rookedsysc.ignitespring.post.domain.usecase.PostDetailUsecase
import org.rookedsysc.ignitespring.post.infrastructure.web.dto.PostDetailDto
import org.rookedsysc.ignitespring.post.mapper.PostMapper
import org.springframework.stereotype.Service

@Service
class PostDetailService(
    private val postInPort: PostInPort,
    private val commentInPort: CommentInPort
) : PostDetailUsecase {
    override fun execute(postId: Long): PostDetailDto {
        val post = postInPort.getPostByIdOrThrow(postId)

        val comments: List<CommentEntity> = commentInPort.findByPostId(postId)
        val commentDtoList: List<CommentPostDetailListDto> = comments.map { comment ->
            CommentMapper.toCommentPostDetailListDto(comment)
        }

        val postDetailDto = PostMapper.toPostDetailDto(post, commentDtoList)
        return postDetailDto
    }
}
